doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,483 @@
+I/O, as (as mentioned above) it would lead to suboptimal performance due to the
+ACK delay issue.
+
+Note that this is orthogonal to whether we provide blocking I/O semantics to the
+application. We can use blocking I/O internally while using this to provide
+either blocking or non-blocking semantics to the application, based on what the
+application requests.
+
+This approach in general requires that a network socket be configured in
+non-blocking mode. Though some OSes support a `MSG_DONTWAIT` flag which allows a
+single I/O operation to be made non-blocking, not all OSes support this (e.g.
+Windows), thus this cannot be relied on. As such, we need to configure any
+socket FD we use into non-blocking mode.
+
+Of the approaches outlined in this document, the use of non-blocking I/O has the
+fewest disadvantages and is the only approach which appears to actually be
+implementable in practice. Moreover, each disadvantage can be readily mitigated:
+
+  - We rely on having a select(3) or poll(3) like function available from the
+    OS.
+
+    However:
+
+    - Firstly, we already rely on select(3) in our code, so this does not appear
+      to raise any portability issues;
+
+    - Secondly, we have the option of providing a custom poller interface which
+      allows an application to provide its own implementation of a
+      select(3)-like function. In fact, this has the potential to be quite
+      powerful and would allow the application to implement its own pollable
+      BIOs, and therefore perform blocking I/O on top of any custom BIO.
+
+      For example, while historically none of our own memory-based BIOs have
+      supported blocking semantics, a sophisticated application could if it
+      wished choose to implement a custom blocking memory BIO and implement a
+      custom poller which synchronises using a custom poll descriptor based
+      around condition variables rather than sockets. Thus this scheme is
+      highly flexible.
+
+      (It is worth noting also that the implementation of blocking semantics at
+      the application level also does not rely on any privileged access to the
+      internals of the QUIC implementation and an application could if it wished
+      build blocking semantics out of a non-blocking QUIC instance; this is not
+      particularly difficult, though providing custom pollers here would mean
+      there should be no need for an application to do so.)
+
+  - Configuring a socket into non-blocking mode might confuse an application.
+
+    However:
+
+    - Applications will already have to make changes to any network-side BIOs,
+      for example switching from a `BIO_s_socket` to a `BIO_s_dgram`, or from a
+      BIO pair to a `BIO_s_dgram_pair`. Custom BIOs will need to be
+      substantially reworked to switch from bytestream semantics to datagram
+      semantics. Such applications will already need substantial changes, and
```

Comment:
As far as I know, QUIC still gives you a byte stream.

Reply:
On the application side, yes. On the network side it uses UDP sockets, and therefore has datagram semantics. `BIO_s_dgram_pair` is for use on the network side, not the application side.